### ISSUE TITLE
tool for MET uncertainties

### ISSUE THREAD

**Issue by veelken** [cms-nanoAOD/nanoAOD-tools#20]
Is there already a plan to compute the systematic uncertainties on the MET (pT and phi), which arise from jet energy scale, jet energy resolution and unclustered energy scale uncertainties ?

I see that there already exist branches in the nanoAOD for computing the unclustered energy uncertainty (MET_MetUnclustEnUpDeltaX, MET_MetUnclustEnUpDeltaY ), but I don't find a tool jet which uses these branches.

**Comment by arizzi** [comment]
from deltaX and deltaY you can comput pt and phi variation

**Comment by veelken** [comment]
Yes, that's what I thought. But we still need a tool that computes the MET systematics arising from JES and JER uncertainties. It would be helpful if this tool applies the deltaX and dletaY for the unclustered energy as well and produces 3 shifted MET (pT and phi) collections.

**Comment by arizzi** [comment]
right. I'm not sure if CMGers already have something ready to import...
otherwise is definitely missing and you can help on that.

Il 20 ott 2017 1:45 PM, "Christian Veelken" <notifications@github.com> ha
scritto:

**Comment by peruzzim** [comment]
For the heppy implementation see in these files:
https://github.com/CERN-PH-CMG/cmg-cmssw/blob/heppy_80X/PhysicsTools/Heppy/python/analyzers/objects/JetAnalyzer.py
https://github.com/CERN-PH-CMG/cmg-cmssw/blob/heppy_80X/PhysicsTools/Heppy/python/physicsutils/JetReCalibrator.py

Currently in CMG three copies of the JetAnalyzer are run, with different values of the shiftJEC configuration parameter.
Please pay attention to the jet selection contained in type1METParams that is used when propagating the JEC shifts to the MET.

**Comment by veelken** [comment]
Hi, 

I am looking into how best to implement the MET uncertainties.

We need to propagate the JES and JER uncertainties to the MET.
The PAT tools which did this is 
    https://github.com/cms-nanoAOD/cmssw/blob/master/PhysicsTools/PatUtils/python/tools/runMETCorrectionsAndUncertainties.py
The tool also smears reconstructed jet energies and propagates the effect of the smearing to the “nominal” MET collection.

Do we apply jet smearing in the nanoAOD production already ?

In case we don’t apply jet smearing yet, I think it will be the best solution to add jet smearing to 
   https://github.com/cms-nanoAOD/nanoAOD-tools/blob/master/python/postprocessing/examples/jecUncertProducer.py
and to produce JES and JER uncertainties for jets simultaneously with the MET systematics, so that the effect of JES and JER uncertainties can be applied in a correlated way between jets and MET on analysis level.

What do you think ?

**Comment by veelken** [comment]
Hi,

I have implemented a first version of the code that computes jet and MET uncertainties in a consistent (correlated) way. You can find the code here:
   https://github.com/veelken/nanoAOD-tools/blob/jetmetUncertainties/python/postprocessing/modules/jme/jetmetUncertainties.py

When running the code via
   python scripts/nano_postproc.py -I PhysicsTools.NanoAODTools.postprocessing.modules.jme.jetmetUncertainties  jetmetUncertainties . $CMSSW_BASE/src/PhysicsTools/NanoAOD/test/nano.root 
I get the output attached below [*].

Does one of you have an idea why the access to the JetResolution class fails ?

Cheers,


Christian

- - - - - - - - -

[*]

Loading jetmetUncertainties from PhysicsTools.NanoAODTools.postprocessing.modules.jme.jetmetUncertainties 
Load Library 'CondFormatsJetMETObjects'
Load Library 'JetMETCorrectionsModules'
Will write selected trees to .
Loading jet energy scale (JES) uncertainties from file '/home/veelken/CMSSW_9_4_0_pre3/src/PhysicsTools/NanoAODTools/data/jme/Summer16_23Sep2016V4_MC_Uncertainty_AK4PFchs.txt'
Loading jet energy resolutions (JER) from file '/home/veelken/CMSSW_9_4_0_pre3/src/PhysicsTools/NanoAODTools/data/jme/Spring16_25nsV10_MC_PtResolution_AK4PFchs.txt'
<ROOT.TClass object ("JetCorrectionUncertainty") at 0x3b65190>
<ROOT.TClass object at 0x(nil)>
Traceback (most recent call last):
  File "scripts/nano_postproc.py", line 46, in <module>
    p.run()
  File "/home/veelken/CMSSW_9_4_0_pre3/python/PhysicsTools/NanoAODTools/postprocessing/framework/postprocessor.py", line 53, in run
    for m in self.modules: m.beginJob()
  File "/home/veelken/CMSSW_9_4_0_pre3/python/PhysicsTools/NanoAODTools/postprocessing/modules/jme/jetmetUncertainties.py", line 113, in beginJob
    self.jer = ROOT.JetResolution(os.path.join(self.jerInputFilePath, self.jerInputFileName))
  File "/cvmfs/cms.cern.ch/slc6_amd64_gcc630/lcg/root/6.10.08/lib/ROOT.py", line 468, in __getattr2
    raise AttributeError( name )
AttributeError: JetResolution

**Comment by veelken** [comment]
I have fixed the error. It seems to have been caused by the fact that a JetResolution class is defined twice (within different namespaces) in CMSSW:
1) CondFormats/JetMETObjects/interface/JetResolution.h
2) JetMETCorrections/Modules/interface/JetResolution.h
and dictionaries were missing.

A first running version of the tool that computes jet and MET uncertainties in a properly correlated way is included in https://github.com/cms-nanoAOD/nanoAOD-tools/pull/24 . Note that this PR requires that the PR #23 is merged first (since PR #24 includes the same change to the btagSF as PR #23).
